### ISSUE TITLE
Add support ,and management for autofill passwords, addresses, etc. (was: edge:// URLs blocked)

### ISSUE THREAD

**Issue by ukandrewc** [MicrosoftEdge/WebView2Feedback#374]
Any way to unblock edge//settings?
If not how do users change stored logins, passwords, etc.
Thanks

[AB#30121282](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/30121282)

**Comment by JoshuaWeber** [comment]
We are actually planning to further change the experience. We are planning to stop storing login / passwords for WebView2. Our thinking on this is to provide for more isolation of experience between an app built on WebView2. And the Edge browser. 

We are evaluating what our long term plans would be for things like login / password auto-fill. Can you share any more details on your app? Is the auto fill a desired behavior? Do you have any other planned mechanism for management of stored data?

**Comment by ukandrewc** [comment]
That is an immense disappointment, it is such an essential part of any browser. Even apps are likely to want user logins.
We implement a complete tabbed browser. So yes it is ESSENTIAL for us. any functionality you disable, we will have to create. It seems an immense shame to have to do that.
I'm sure we're not the only ones, so please don't make us redo all that work, that already exists.
I'm not quite sure why you would decide to limit it's functionality. Surely better to allow it to be disabled, rather than removing a complex and essentail part of the browser altogether?

**Comment by ukandrewc** [comment]
Can I ask if you are palling to remove any other functionality needed by a browser?
We currently use GeckoFX, but need to move to a more modern browser.
We were hoping that WebView2 would fulfil our expectations, but it sounds like you don't want it to be used as a browser?

**Comment by JoshuaWeber** [comment]
We for sure are happy if you use it to build your own browser. But we expect that is not the primary use case of WebView2. And our principle and goal is to support all browser functionality in the WebView2 context. 

We do have some legal and product needs to maintain some independence between the Edge browser and WebView2. This was a big part of our motivation around blocking the edge://settings page. And in our restriction on the auto-fill of logins and passwords. 

That said, we would like to bring this functionality back. In a more dev controlled manner. If we did support auto-fill for things like passwords. What would be your ideal experience. As an example, would you expect to render the settings and management UI yourself and take advantage of a WebView2 provide API for passwords? Or would you expect built in WebView2 UI for that functionality (similar to what you had before at edge://settings)? Or another option?

**Comment by douglas-jordan** [comment]
@JoshuaWeber in our application we automate URLs so the customer can quickly open many different pages.  These pages can be the same ones they would use in a desktop browser.  The user experience should be easily identifiable for things like username/passwords entry and not so different they would not know what to do.  Default look and feel for what the desktop browser offers is sufficient for most needs in this area of use.

**Comment by ukandrewc** [comment]
That's great, thanks for the clarification. The hardest part to code, is the UI experience, so as you say, an autofill API, where devs provide the settings and management, would be excellent.
BTW, thanks for your open approach to the development of this project.

**Comment by billybooth** [comment]
@JoshuaWeber said:
> We are planning to stop storing login / passwords for WebView2.

Just want to add my two cents here.  We intend to use `WebView2` for web-based authentication against an upstream/identity provider.  Credential auto-fill is absolutely essential for us for subsequent logins, although the ability to the feature on and off would certainly be nice.  Since we are not creating a full browser experience, the distinction between exposing a password API vs a built-in equivalent to `edge://settings` is not important, but auto-fill is absolutely essential.  Thanks for your consideration!

**Comment by billybooth** [comment]
@JoshuaWeber, as I see it, our product will be using the `WebView2` in a couple of ways.  Initially, for authentication via a third-party identity provider (think SAML SSO).  In this case, whether the login data is preserved/auto-filled on subsequent product launches will be dictated by an application-level configuration.  The user does not need tools to manage the auto-fill data, only perhaps the ability to clear the entire auto-fill cache (via an API that our product would interact with).  The ability to clear the auto-fill data can currently be done by simply trashing the `WebView2`'s data directory, although that's obviously not ideal, since we throw the baby out with the bath water and don't want to go digging through the filesystem.

Were I able to choose the "ideal" experience for our product and our users, it would be that the `CoreWebView2` could be configured to use one of the following policies, where `1)` is what is currently provided:
1) Use Autofill (and let the user choose to whether to save credentials via built-in prompting)
2) Use Autofill (and automatically save credentials)
3) Disable Autofill

Then additionally, for us, only a simple management API is required with the ability to clear all the auto-fill data.  Keep in mind I am speaking only for myself, I understand that in the context of a fuller desktop browser implementation, a far more robust management API or built-in interface is required.

The second use-case within our product is for something more akin to a traditional desktop browser view, but again, I believe the extent of our needs is the ability to dictate a) whether Autofill will be used (bonus point for how it will work), and b) the ability to clear auto-fill data programmatically.

**Comment by JoshuaWeber** [comment]
At GA autofill will be removed. It will be a post GA option for us to re-enable it. It is very high on our backlog for a post GA item.

I'm going to try to work to get a re-design proposal out quickly. So we can get some fast feedback from you and get it built and re-added sooner.

Auto-fill has 3 parts.
1) passwords
2) Form filling
3) Payments

Can you let me know which of the 3 you care about. And provide a ranking for which you care about most?

**Comment by ukandrewc** [comment]
@JoshuaWeber Thanks, for us it would be only be usernames and passwords. Our app browses to admin areas on shopping carts. Being able to autofill logins for those, is really essential.

**Comment by billybooth** [comment]
Probably apparent from the details provided above, but credential autofill is all that is needed for us as well.

**Comment by champnic** [comment]
I'm going to reopen this issue to keep it tracking the ask for autofill management. I've linked it to our backlog item. Thanks!

**Comment by larsandberg** [comment]
@JoshuaWeber What is the current status on this? You mentioned the possiblity of a "_... post GA option for us to re-enable it_". We are in a desperate need to be able to offer autofilled passwords! Otherwise we cannot launch WebView2 as a part of our planned release in february 2021.

**Comment by ukandrewc** [comment]
@larsandberg If it helps, there are JS libraries that can do this, I've been looking at this: https://tarekraafat.github.io/autoComplete.js/demo/

**Comment by larsandberg** [comment]
@ukandrewc Thanks for your suggestion! Does that also support autofilling of stored usernames & pwd:s

**Comment by ukandrewc** [comment]
Not directly, you'd need to manage that. I am starting work on one here: https://github.com/ukandrewc/Webview2.Winforms.DOM but I may drop it, if the MS version is ready in time.

**Comment by larsandberg** [comment]
@champnic What is the current status on this? You mentioned the possiblity of a "... post GA option for us to re-enable it". We are in a desperate need to be able to offer autofilled passwords! Otherwise we cannot launch WebView2 as a part of our planned release in february 2021.

**Comment by ukandrewc** [comment]
@champnic  Just another vote for a timeline on this please. Just need to know if I should start coding my own. Thanks.

**Comment by hunkydoryrepair** [comment]
If this was disabled in GA was it enabled again for 1.0.721-prerelease? I am seeing autofill on all fields in our app when using this version.
It's actually a major problem because we are used in financial sector, entering customer personal information in fields, and then to see it pop up with the wrong customer name or phone or social security number is a big security problem.  And no browser seems to respect the autocomplete="off"

**Comment by maurawinstanley** [comment]
@hunkydoryrepair Thanks for letting us know, we do have autofill disabled for passwords and web payments, and are tracking a scenario to disable the rest of autofill.  We will update this thread with more information soon.

**Comment by hunkydoryrepair** [comment]
It's a rather large assumption on behalf of browser developers that an input field will accept the same values every time it is used.  And when we have javascript to handle our own autocomplete like features, the browser autocomplete dropdown covers up the javascript implementation.

**Comment by maurawinstanley** [comment]
@hunkydoryrepair sorry for the delay in our response, the majority of the team is on holiday and we will get back to you early next week.  Thanks again for brining this to our attention!

**Comment by jasonstephen15** [comment]
Hi all, as part of enabling/disabling autofill we want to give developers a way to manage cached autofill data, but unfortunately can't simply expose the edge settings:// page as Josh mentioned above. 

We are considering releasing an API to clear all cache (not just autofill). Would this API suffice for your app's purposes? Or is more granularity needed, and if so why? Trying to gather data points, so any input would be great! Thanks.

**Comment by ukandrewc** [comment]
@jasonstephen15 I do suspect that it would need to be more granular, but what else would be cleared?

**Comment by hunkydoryrepair** [comment]
Personally, I'm not looking to manage it, I just want to turn it off, like a kiosk mode so even information from the same session or even same page load should never be displayed.   It's a very greedy UI and we want new values every time. Currently, we try to outsmart Edge by inserting hidden input fields, but it's a battle as Edge keeps trying to undo any attempt to turn this off. 

But, clearing the entire cache seems extreme when there may be 10MB+ of javascript and css. Also, we make heavy use of localStorage for preferences  and we need to make sure that never gets cleared.

**Comment by billybooth** [comment]
@jasonstephen15 I agree with the comments above.  In addition to the ability of the "autofill" API to completely enable/disable the autofill functionality, the ability to clear only the autofill portion of the cache (without trashing the entire cache) seems necessary.

I suspect others will want far more granularity than this and a full-fledged management API, but the above feels like an acceptable starting point.

Personally, a big one for us is the ability to manage the (credential) autofill prompting; as I said above, we would like to be able to manage whether credentials are automatically saved or whether the user is prompted to save credentials.

**Comment by jasonstephen15** [comment]
@ukandrewc - I believe clearing cache would essentially do the same thing as dumping the user data folder, but would be able to be called at runtime.

**Comment by jasonstephen15** [comment]
@billybooth - Thanks for the input! Enabling password autofill, should take care of your use case, which we are planning to do. We intend to still have a password saving prompt show up, as it does in Edge, but I suspect there might be a way either through intercepting the popup, or through javascript that could automatically save credentials.

**Comment by ukandrewc** [comment]
@jasonstephen15 OK, thanks. That wouldn't be any use to us, we want to keep and manage login details, not just delete them.

**Comment by avinashtauro-arctechinfo** [comment]
Our specific requirement was to disable autofill completely as our UI running WebView2 would run all day on kiosk mode.
Multiple users would access the kiosk and we didn't want the next user to see the autofill options of the previous user.

This following worked for us as a workaround till this feature to turn off autofill is available.

1. On Page Load Change the name of the text box to a random name. (This turns off autofill)
    - randomizeInputNames(false);
2. If you need to POST the values in the text boxes, dynamically create hidden boxes with the same name as the original textboxes, and update the hidden value with the textbox value, in onchange or onpost.
    - randomizeInputNames(**false**);

//AutofillWorkaround.js
var randomPostFix = getRandomFileNamePostFix();

function randomizeInputNames(supportFormPost) {
    var inputElements = document.getElementsByClassName('randomize');

    // iterate all elements with class = randomize
    // assuming only input elements
    for (var i in inputElements) {
        var input = inputElements[i];

        // Backup the textbox name before changing it
        var inputName = input.name;

        // Rename textbox name to a standard random name
        input.name = inputName + randomPostFix;

        // If the textbox data is to be POST'ed then we have to 
        // 1. create hidden fields with the same name and value as the original textbox
        // 2. Add a onchange event handler to the textbox to update the hidden element Value
        if (supportFormPost) {
            // The below two function are only needed if the input elements need to be POST'ed
            createHiddenField(inputName, input.value);
            input.onchange = onTextChange;
        }
    }
}

function getRandomFileNamePostFix() {
    var date = new Date();
    var fileNamePostFix = "_" + date.getFullYear() + "_" + (date.getMonth() + 1) + "_" + date.getDate() + "_" +
        date.getHours() + "_" + date.getMinutes() + "_" + date.getSeconds();

    return fileNamePostFix;
}

// The below two function are only needed if the input elements need to be POST'ed
function createHiddenField(name, value) {
    var input = document.createElement("input");

    input.setAttribute("type", "hidden");
    input.setAttribute("name", name);
    input.setAttribute("value", value);
    input.id = name;

    var form = document.forms[0];
    form.appendChild(input);
}

function onTextChange() {
    var hidId = this.name.replace(randomPostFix, "");
    var hidElement = document.getElementById(hidId);

    hidElement.value = this.value;
};

**Comment by jasonstephen15** [comment]
Hi all, thank you for your patience. We have made progress on the Autofill API. We have been working with the Edge team to understand the nuance/edge cases with autofill. 

One concern we still have outstanding is regarding 'autostuffing' - this is a functionality in edge that populates a username and password field immediately when you enter a site (and have cached password data). Currently, when a user disables password-autofill in the browser, autostuffing still exists. Meaning if you had cached password data, the username/password box would autopopulate, and have a dropdown, even though you have disabled password-autofill. WebView doesn't have an easy way to disable this, and we'd like to know your thoughts on this.

We are thinking about shipping a clearing cache api along with this. So one option would be to clear cache every toggle, to ensure there is no password cache when autofill is off, therefore no autostuffing.

What I'm seeking to validate is if many people want to toggle autofill on/off, or simply set it once for your app? If all that is requested is to set autofill on/off once for your app, the API should work for your use case.

Would appreciate any feedback/thoughts here! Thank you!

**Comment by monkeypro** [comment]
We would like to have password autofill (autostuffing) always ON. 
We will not toggle on and off

**Comment by ukandrewc** [comment]
@jasonstephen15 Not sure I have got this: Is this clearing a cache for the current page or all autofill data?

**Comment by billybooth** [comment]
> Not sure I have got this: Is this clearing a cache for the current page or all autofill data?

I'm assuming this means that an exposed cache API would clear all autofill data when toggling autofill off.  Clearing it for the current page wouldn't seem to do much good if there were the potential to navigate to another page that may also have autofill data cached.

@jasonstephen15, We would definitely need the ability to toggle autofill on and off, since the behavior will be managed by a setting in our own application and may be changed by a user.

**Comment by ukandrewc** [comment]
@billybooth I assumed the same, just wanted to clarify.

As long as the autofill data was still available when it's turned back on, that will be OK for us.

**Comment by jasonstephen15** [comment]
@ukandrewc @billybooth - Got it, just to clarify - toggling on and off is definitely doable. The bit of confusion is with autostuffing still happening if there is cached password data AND you toggle passwords-autofill off. This is the current functionality in Edge.

If you'd like to repro:
1. On Edge, save a password. 
2. Disable passwords saving in edge settings. 
3. Navigate to the site you saved your password for. You will notice the password still autostuffs.

This works in edge, bc there is an easy way to manage passwords through UI in the address bar. And we don't have that in WebView2.

Does this seem like a confusion/problem for your end users?

**Comment by billybooth** [comment]
Our use case (the one that depends on password auto-fill) is very narrow.  It will present the user with an SSO login page upon each launch of the application and again if the active session times out.  If the preferences are changed to disable password auto-fill, we will need to clear the auto-fill cache, because the same page will be displayed the next time the user is required to authenticate.  If the prior username and password are autostuffed, it will not only be confusing to the user, but it will represent a fairly significant security issue in shared environments.

**Comment by darkguy2008** [comment]
Well, I totally agree with the fact that the idea of the WebView2 control is to **not** be a drop-down replacement for a full-fledged browser. I mean, if we had a lot of custom browser apps that used the same control then they all would behave the same way so if there's an exploit for one, it could exist for all the others, so from a security perspective I don't think that's a good idea, either. Plus, the idea (if I'm not mistaken) of the WebView2 control is to be a way to use the newest **rendering _engine_** where you would use the standard WebView or other different web controls inside an app.

Having set that aside, in my use-case I'm making a desktop app using WebView2 as the rendering engine, with a custom UI of course. So the autofill dialog is alien to the UI, and in my app we don't want that to happen, either. IMO that's for a browser, not for a desktop app. So, if it's too hard to remove the autofill setting (which is understandable considering how annoying Chrome stuff can be), if there's a way to programmatically clear the cache (so the dialog or autostuffing don't happen) then **yes**, definitely do that please!

I'm all for giving the programmers/developers/users of your app/library/toolkit/framework/etc. **the freedom to choose/toggle features**. It would be a complete turndown if Microsoft decided that it wasn't going to do that, considering that flexibility is one of the things it's also looking for in this new WebView2 control, and not letting an user control a specific feature (or at least provide a workaround) definitely goes against that.

So my vote goes to either allowing us to toggle autofill/autostuff, or an easy way to clear the cache so it doesn't happen. Those who want it allowed 100% should also think on the ones that may not want or need to have that feature 100% enabled.

**Comment by ansalc** [comment]
I, at the moment, often enter login credentials injecting keyboard inputs. I would very much like to have full autofill functionality like in Chrome or Edge.

**Comment by billybooth** [comment]
Just checking in on this issue and again lobbying for suggestions/creative solutions for a problem I'm anticipating having.

We're using the `WebView2` for an authentication dialog that displays the login page of an upstream identity provider.  When the user authenticates, if MFA is not enabled for the account, there is opportunity for the user to see or accept the save credentials prompt, because we're immediately redirected to a SAML POST request.

The ideal solution for this problem would be a `WebView2` configuration that allows autofill credentials to be saved with the standard prompt, automatically, or not at all.  But if I don't get that option, am I completely out of luck?

**Comment by hunkydoryrepair** [comment]
I still can't ship with the new stable release as it still has dead links to edge pages pop up for managing identities.  I thought this was turned off? Are there any settings we control as of 1.0.774.44? command line switches or anything to just turn this off so we eliminate dead links and can actually work in a kiosk mode?

**Comment by rhickmott** [comment]
This pretty much makes the project pointless.

Seeing as Microsoft have chosen to ignore autocomplete settings on forms and items (why ffs) there is no way to access the flag to shut out the option to disable this feature and every text field comes with a garish auto fill prompt without a lot of sodding around on the backend.

Kiosk apps which is a major use of this and exposes previous users data on the form as well as giving them a way to break out of walled apps by launching a edge settings window complete with address bar (brilliant)

**Comment by monkeypro** [comment]
Have anyone tried to turn **on** password management by using the workaround from this issue?
[Add ability to disable saving of password](https://github.com/MicrosoftEdge/WebView2Feedback/issues/298)

**Comment by rhickmott** [comment]
I haven't turned it on however we do use a similar workaround to disable Auto Fill (in absence of a better way)

The file can be parsed using a Json Parser and I would guess you just need to work out what property to turn "on" and then edit the file accordingly. If it works you can just parse the file, change/add the flags and save back.

**Comment by champnic** [comment]
Hey @rhickmott - sorry you feel this way. We haven't chosen to ignore the settings, and are currently working on this item to provide a setting to control the autofill features. The link to edge settings has already been fixed and is available in runtime versions 90.0.787.0+. You should be able to test this in Microsoft Edge Canary and Dev channels, and it will ship with the next stable WebView2 Runtime release, 90.0.818.x.

**Comment by rhickmott** [comment]
Hi @champnic by ignore I wasn't referring to the Dev Team ignoring the settings but to Microsofts Direct Choice (in Edge) to ignore the autocomplete="off" flag placed on HTML Field Elements and Forms and provide one anyway 🤦‍♂️ thankfully I have managed to hack a way round it but do look forward to better API control when it hits.

I'm not going to debate whether this is a good idea or not as it's outside the scope of this problem, however the woes are there if you search the internet along with the hacky JavaScript solutions to provide "unique" names to fields and to preserve them on post 🤢

For WebView a large use case for this is for Kiosk or Walled Browser Solutions and going against the wishes of the page developer and adding in autofills when the developer doesen't want them with no way to disable them is a large problem.

**Comment by monkeypro** [comment]
Hi @champnic 
Can you confirm that it should br possible to turn ON password manager by using the workaround above?

**Comment by champnic** [comment]
@rhickmott Ah gotcha, thanks for the clarification. Sorry you had to hack around it - I did some digging and I think we are matching chromium's behavior here (for now). This bug doesn't look encouraging: https://bugs.chromium.org/p/chromium/issues/detail?id=587466
If you are seeing a difference in behavior from Chrome/Chromium please let me know.

@monkeypro I haven't tried that workaround so can't confirm it myself. The API we are developing will have the ability to enable password autofill.

**Comment by markismail7** [comment]
So for I'm loving Webview2. We've needed it 10 years ago. Its finally here. AutoFill Password is necessary to the user. I've already gotten multiple requests for that. Let the user decide to use the feature or not. Thanks

**Comment by darkguy2008** [comment]
> So for I'm loving Webview2. We've needed it 10 years ago. Its finally here. AutoFill Password is necessary to the user. I've already gotten multiple requests for that. Let the user decide to use the feature or not. Thanks

That's cool if your users would want it. Mine don't. Would you want your password manager app (i.e. NordPass, which uses a similar UI) to save autofill regardless of your preferences so anyone can enter and access all your passwords? This is just an example.

I'm all cool with letting the users to decide whether if they want autofill or not, but we, as **developers** we should be able to decide if we want to allow that feature to be toggled or not, instead of resorting to ugly hacks like it's been stated before.

**Comment by markismail7** [comment]
Either or. We can manage the passwords ourselves if that's a problem. I agree with you on giving the developer the option to choose. Because I've noticed some need it badly and some do not want it at all.

**Comment by champnic** [comment]
Agreed - the current plan is to give devs the choice on using the autofill password behavior or not through a WebView2 setting.

**Comment by barrynovak5** [comment]
We want to use WebView2 on web sites that require HIPAA-compliance (i.e. protecting patient data). So disabling auto-fill is critically necessary. As other developers have commented, give us the ability to control the pieces individually (enabling/disabling caching of passwords, form data, payments). Is there a target date for when this feature will be available?

**Comment by billybooth** [comment]
> @champnic said:
> [...] the current plan is to give devs the choice on using the autofill password behavior or not through a WebView2 setting.

@champnic, I apologize for belaboring this, but I'm still hoping to get some clarity about whether the to-be-provided autofill setting can be represented by a policy enum rather than a boolean to leave the door open to something like: `WithPrompt`/`Automatic`/`Disabled` for the credential autofill policy.

If there isn't a snowball's chance, just let me know and I'll move along. 😉

**Comment by champnic** [comment]
@barrynovak5 No specific date, but it will probably be available as an experimental API in the next prerelease SDK package (around mid-April).

@billybooth Sorry I missed your previous question! Right now the design is to either have the current behavior (the prompt, which sounds like it is insufficient in your scenario) or disabled. I'll make sure the team sees this feedback. Even if it doesn't become an enum, it makes sense to potentially have this be configurable if credential autofill is enabled (WithPrompt or Automatic) and could be a separate but related setting.

**Comment by jasonstephen15** [comment]
@billybooth - Apologize for the delay. Trying to understand your request a little better. Our current plan is for password autofill to be disabled by default, and with our API release, devs will have the option to enable the the default edge (minus some branding) password saving prompt. (Which hasn't yet been available in webview2). This should take care of your request of having WithPrompt and Disabled options. 

Regarding 'Automatic', I think there is some security concern here in allowing devs to programmatically save the end-users password, without prompting the user first. This could be used maliciously and my first instinct is to say we probably would not support that, but I am open to discussion & other opinions.

Regarding your current state - I’m not sure I understand fully why the your workflow doesn't allow saving passwords - are you creating your own password-saving dialogue? The edge/webview2 password prompt we intend to implement is not light dismiss, so it should persist past the login page, unless you are closing the webview immediately after the user enters their password for some reason. If this is the case, then I'd imagine its a timing issue? Perhaps its possible to code a time delay to give their end-user a chance to save password, before closing. 

Either way, I think our upcoming solution should meet your needs (barring the Automatic option). Please let me know if this is not the case. Thanks!

**Comment by darbid** [comment]
> Regarding 'Automatic', I think there is some security concern here in allowing devs to programmatically save the end-users password, without prompting the user first. This could be used maliciously and my first instinct is to say we probably would not support that, but I am open to discussion & other opinions.

@jasonstephen15 this is a valid argument and certainly a huge topic with varying opinions. I would like to suggest that WebView2 is not a browser.  Please leave the security topics, choices to the developer of the browser/application. At the end of the day it is the developer of the Browser/Application that answers to the users, supplies licenses and privacy statements.

Apps like 1Password have extensions to Chrome/Edge and use Javascript I suspect. So at the end of the day if the reason for blocking "Automatic" is a security concern then it will just mean a little more work.

**Comment by jasonstephen15** [comment]
Thanks @darbid & @billybooth for the perspectives, you both bring up valid points. Barring the privacy concern, theres definitely value in the ability to save credentials automatically via the webview2 password prompt for end users instead potentially creating a JavaScript workaround. What I can do is double check with our privacy/legal team, and make sure we are good on the here. Will follow up shortly!

@billybooth - To confirm, the automatically saving option is a blocking issue for you, correct? There is no way to create a custom prompt for the user after their credentials are authenticated from the upstream IP? In this case, would saving the password automatically, even be useful? Because the saving process would probably occur before the credentials are verified from IP (the user could be autosaving the wrong password). Additionally this raises the question of what to do if the password does save incorrectly, or is changed by the IP.

Reason I'm asking is because privacy concerns aside, I'm trying to gauge if this is a common request where end-users are not given much chance to be prompted to save their password. Are there other use cases where automatic programmatic password saving is required? Thanks again for the time and feedback!

**Comment by darkguy2008** [comment]
@jasonstephen15 

I'd like to chip in a bit because this is an issue that I foresee is going to affect me in the near future in an app I'm developing, and other cases, too:

Sometimes, depending on the app, you might **not** want autofill to happen. As I stated in past posts, for apps like NordPass or any other password management app, heck, even a game, maybe, say, League of Legends launcher), having autofill enabled is a big security concern, as it could have been run in a shared computer or somewhere and if the developer can't turn off autofill, a malicious user could just guess the email and have the control autofill the password (or autofill the email and harvest them, too). Not a good thing.

Regarding autofill per se, considering that this issue ( https://stackoverflow.com/questions/15738259/disabling-chrome-autofill ) was asked 8 years ago and is still relevant, and that I've also chimed in in Chromium's bug reports ( https://bugs.chromium.org/p/chromium/issues/detail?id=914451 ), and that a bug like this with 63 comments so far is also happening... Well, if devs have to go through all these hoops and find workarounds, then it's evidently bad design. I cross my fingers so the same doesn't happen to WebView2 (Taking @champnic's words, it seems the team is on the right track, good to know!).

I also think you/we have to be very careful here for what you/we want the WebView2 control to be and what you/we expect. If I'm not wrong, the idea is for it to be a replacement / improvement over the old WebView control, and as a web browser control, it should **not** have any kind of extra UI elements, screens or pages that could differ or create a different UX from what the app developers intend.

As for the app I foresee this issue might (and hopefully not) affect, my startup's app (I signed an NDA though, so can't say much) is to be used on shared computers, so the less I'd want is the lack of liability to correctly manage and store my users data. If other people would want to access a previous user's email/account, if the autofill setting is enabled (or worse, what if somebody discovers an exploit inside WebView2's autofill code that we (currently) can't control?) then the results could be catastrophic.

I 100% agree with placing that responsibility on the developer. Who, also, bear in mind, receives instructions from either security/compliance teams and UX gurus who, for example, would like the app to **not** store autofill data or even show hints to it or leave open doors to manage them if they exist for security reasons, or the UX team wants a consistent UI look and feel and if we can't control those settings "from the outside" then we're left with whatever the WebView2 team decides in terms of UI, looks and usability, which may not align to the UX team's expectations.

So, to summarize, the less workarounds we as devs have to go through and the more options we have to use an external asset, the better experience for devs and users :)

**Comment by billybooth** [comment]
> To confirm, the automatically saving option is a blocking issue for you, correct? 

I'll be honest, as much we want this feature, it is not going to be a blocker for us. We will be able to move forward with the autofill options you can provide, even if they don't fully meet our needs.

> There is no way to create a custom prompt for the user after their credentials are authenticated from the upstream IP? 

When using a prompt to save autofill credentials, we will use the built-in prompt after authentication by swallowing the last redirect, scraping what we need, and leaving the authentication dialog open for a limited window of time to allow the user to accept the built-in WebView2 prompt.

> In this case, would saving the password automatically, even be useful? Because the saving process would probably occur before the credentials are verified from IP (meaning the user could be autosaving the wrong password).

Yes, this is what we want. We know whether the user authenticates successfully because we will either have a valid authorization token or not. If the user provides incorrect credentials and they are retained, so be it. The user will eventually provide the correct credentials or not be able to launch the application.

> Additionally this raises the question of what to do if the password does auto save incorrectly, or is changed by the IP. 

If it is changed by the IP, the situation is no different than if autofill credentials are saved manually via the prompt. If incorrect credentials are saved, we can live with that. The nature of automatic credential retention is that the last utilized credentials are retained, and we accept that.

> WebView2 doesn't offer a password management API yet. Any thoughts on how to handle this case?

We know whether the user authenticates successfully or not, and in the future, we will leverage the management API to remove any automatically retained credentials after a failed login. In the meantime, we think it is acceptable to retain invalid credentials.

Thanks so much for your engagement on this request. I'll get back to you tomorrow with my thoughts about other use-cases for this functionality.

**Comment by trogmaniac** [comment]
Just wanted to add my +1 to this thread, it's an important feature to have.

> @barrynovak5 No specific date, but it will probably be available as an experimental API in the next prerelease SDK package (around mid-April).
> 
Are we close to a release, or will there be a delay?

At which point in time could this feature make it into the "Evergreen" installer at the earliest?

**Comment by champnic** [comment]
It will be available in the next pre-release SDK, which will go out this week or early next. When it does, it will be an experimental feature and will work with the Canary and Dev Microsoft Edge runtimes. It won't be in the stable evergreen WebView2 Runtime until May or June.

**Comment by hunkydoryrepair** [comment]
Does this problem just seem to be getting worse? At first, it was only happening for me on address fields, but now all text fields are saving entries and suggesting data, and providing the message "type "/" to search your saved data".

So while I worked around it on some address fields, it has become hopeless to stay ahead of it. Leaking private information everywhere.  This fix cannot come too soon.

**Comment by maurawinstanley** [comment]
We are working on a Clear Browsing Data API in which developers can pass the browsing data type they would like to clear.  This would help with autofill/password form clearing.  We are interested in hearing if adding a time parameter is something that would be of use.  Currently the API intention is to clear that browsing data type for all time and the endpoint looks like: 

HRESULT ClearBrowsingData(
    [in] uint64 dataKinds,
    [in] ICoreWebView2ClearBrowsingDataCompletedHandler *handler);
    
If the time parameter seems necessary/useful the API would clear data between the start and end time.  Our hope is that this would give a bit more flexibility on the data to clear and would look like this:

HRESULT ClearBrowsingData(
    [in] uint64 dataKinds,
    [in] double startTime, [in] double endTime,
    [in] ICoreWebView2ClearBrowsingDataCompletedHandler *handler);
  
The time parameters are the number of seconds since the UNIX epoch and the dataKinds will be one or more data types to clear.

**Comment by maurawinstanley** [comment]
We are planning on providing support for these data types (subject to change): 
* [DOM storage]  = app cache, file systems, indexed DB, local storage, web SQL, and cache storage (each of the listed data types will be individually supported as well)
* [Site data] = DOM storage, cookies (each of the listed data types will be individually supported as well)
* Download history
* General autofill forms
* Password autosave forms
* Browsing history
* Settings
* HTTP Cache

The dataKinds enum will be a mask composed of one or more of the above data types.

**Comment by ukandrewc** [comment]
@maurawinstanley Thanks, that looks good. As you know, really need the ability to edit as well as clear autofill entries.

**Comment by cglacken** [comment]
@maurawinstanley any update on this one? I do not see a milestone set and we are looking forward to this one as well.

**Comment by MysteriousGreenBean** [comment]
Will this feature also allow access to NTLM credentials?
![image](https://user-images.githubusercontent.com/12065508/118811464-56337d00-b8ad-11eb-930a-1a078f3a03d7.png)

It would be really useful to have the ability to provide those data for the user from application itself.

**Comment by champnic** [comment]
@MysteriousGreenBean We are working on an authentication API in #120 that I believe would include this scenario and allow your app to provide the credentials.

**Comment by ryan-veracitek** [comment]
+1 ... really need the ability to enable saving login info. Since COVID many users access the system from a computer that's off the domain & they're having to retype auth everytime they load. Thank you.

**Comment by rhickmott** [comment]
Are we any further on with this?

The method we were using to disable this (setting the json setting in the profile to false) apparently no longer works as something is setting it back to true.

**Comment by billybooth** [comment]
Hoping we finally see something in the upcoming pre-release.  We're getting close to shipping and will obviously be without this functionality for our upcoming GA, but hopefully we'll get it in the subsequent WebView release a month-and-a-half from now.  Any tidbits for us?

**Comment by champnic** [comment]
Hey all - the password and general autofill settings were just promoted as stable APIs in the upcoming prerelease SDK (available this week or next) and they'll be available in the following release SDK. Thanks!

**Comment by champnic** [comment]
The stable API is available in 1.0.865-prerelease SDK, and will be available in the following release SDK. Closing this issue for now. Thanks everyone for the engagement and discussion here! We have a follow-up item to enable detailed control over autofill entries in #1261.

**Comment by ansalc** [comment]
Great. And how do we use it? Is there any documentation about it or at least, can someone point to the methods available to  manage saved autofill/passwords?

**Comment by ukandrewc** [comment]
Other than the name change of `Settings.IsPasswordAutosaveEnabled`, there's nothing else in 1.0.902.0 ;-(

**Comment by billybooth** [comment]
> Other than the name change of `Settings.IsPasswordAutosaveEnabled`, there's nothing else in 1.0.902.0 ;-(

I haven't tried yet, but the documentation seems to suggest that this is exactly it, with a simple autofill cache management API at this point not being present at all in `1.0.902-prerelease`.  Since we're using this for a very limited purpose at this point, setting this property to `true` and then dropping all existing WebView2 UserData when toggling it to `false` sounds like it probably covers our simplest on/off autofill requirement.  Seems like the short-term simple autofill cache management API that seemed like it was coming was dropped in favor of providing the full API covered by #1261 (at some point).

https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2settings.ispasswordautosaveenabled?view=webview2-dotnet-1.0.902-prerelease

**Comment by ukandrewc** [comment]
@billybooth Thanks for the clarification, I was thrown by a single setting being called an API ;-(

**Comment by monkeypro** [comment]
Is there anyway to check if the _Save password_ dialog is open? 
After a sucessful login we close the window containing webview2.
But if the _Save password_ dialog is open we need to wait until the user have answered the dialog

**Comment by billybooth** [comment]
@monkeypro said:
> Is there anyway to check if the _Save password_ dialog is open?
> After a sucessful login we close the window containing webview2.
> But if the _Save password_ dialog is open we need to wait until the user have answered the dialog

We have a similar situation whereby we dismiss the WebView2 form after successful authentication, but we still want to support credential save / autofill.  I have been lobbying for a policy option to automatically save credentials for scenarios like this.  At present, there is no mechanism I'm aware of to accomplish this other than to leave our authentication dialog open for an arbitrary amount of time to allow the user to save the credentials. Obviously undesirable.  Glad to hear others have a similar need.

**Comment by champnic** [comment]
We are currently working on the cache clearing API (part of #140): https://github.com/MicrosoftEdge/WebView2Feedback/pull/1168/files?short_path=55746ae#diff-55746aeb7a0e49f13e6f4df784db43edbb58caf36ded32e565549075042e6528

**Comment by billybooth** [comment]
Anybody else having issues with the current Canary Channel build of Edge?

It seems a bit like they recently moved things around (i.e., from `Program Files (x86)` to local `AppData`) and the loader is no longer able to resolve the webview2 runtime from the canary builds.  I tried with both the x64 and x86 Canary builds, but no dice:
![image](https://user-images.githubusercontent.com/20958170/121069814-8f1b8f00-c793-11eb-9f38-276d0ea3e65d.png)

I haven't run preview in a while, so maybe I'm just doing something stupid, but if others are experiencing the same, I'll file an issue.

**Comment by maurawinstanley** [comment]
@cglacken Sorry about the delay in the update, we are hoping that it will make it into the next pre-release as an experimental feature.  We had a couple of last-minute edge-cases that we are dealing with.  Thanks for your patience!

**Comment by hunkydoryrepair** [comment]
> @billybooth Thanks for the clarification, I was thrown by a single setting being called an API ;-(

Well, there is also `IsGeneralAutofillEnabled`, so that's 2 settings.

**Comment by ukandrewc** [comment]
@hunkydoryrepair but you seem to be forgetting the 17th century adage: 'Two properties do not an API make'

**Comment by champnic** [comment]
@ukandrewc True, and they clarified in the 18th c. - "Two's a couple, three's an API"

**Comment by markismail7** [comment]
I have a Click once application that uses webview2. When a user login for example to Gmail, webview2 keeps Gmail logged in. I love that. However, after updating the application, everything resets. And the problem is:  we do a lot of updates. Is there any way to keep webview2 from not resetting after updating click once application?

**Comment by champnic** [comment]
@markismail7 When you update the application, are you doing anything to the User Data Folder? That is where the cookie is stored that would keeps folks logged in, so if it's getting deleted during an update that would exhibit this behavior:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/user-data-folder

**Comment by rhickmott** [comment]
Are they session cookies or cookies that are set with a expiry date.

I noticed that session cookies that are normally deleted when the browser is closed don’t get cleared with Webview and stick around between app launches. I don’t know if this is intended or not.

Can’t vouch whether they stick around in upgrade though 😉

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by champnic** [comment]
@rhickmott That sounds like it may be a bug on our end if cookies that should be getting cleared are not. Did you open a separate issue for that?

**Comment by CoastalData** [comment]
> I have a Click once application that uses webview2. When a user login for example to Gmail, webview2 keeps Gmail logged in. I love that. However, after updating the application, everything resets. And the problem is: we do a lot of updates. Is there any way to keep webview2 from not resetting after updating click once application?

I've noticed this behavior as well! It doesn't matter the change, when you publish, any site you were logged in to with webview2 will now be logged off. 

Interestingly, I run the development binary from within the Visual Studio project, and the settings do not get lost; it is _not_ the changes to program that cause this behavior, it is definitely something to do with how it gets published via Click once.

**Comment by AlbertoCe** [comment]
> @monkeypro said:
> 
> > Is there anyway to check if the _Save password_ dialog is open?
> > After a sucessful login we close the window containing webview2.
> > But if the _Save password_ dialog is open we need to wait until the user have answered the dialog
> 
> We have a similar situation whereby we dismiss the WebView2 form after successful authentication, but we still want to support credential save / autofill. I have been lobbying for a policy option to automatically save credentials for scenarios like this. At present, there is no mechanism I'm aware of to accomplish this other than to leave our authentication dialog open for an arbitrary amount of time to allow the user to save the credentials. Obviously undesirable. Glad to hear others have a similar need.

Hi @monkeypro and @billybooth,
Have you found any way to accomplish this ?

**Comment by monkeypro** [comment]
@AlbertoCe 
We just leave our authentication dialog open for a short time to allow the user to save the credentials. 😕
